### PR TITLE
feat(report): #MZI-76 send delivery report

### PR DESCRIPTION
**Pull request by qpic-cgi**
## Describe your changes
Send delivery report to original sender when it's necessary

## Checklist tests
- with "user1", from zimbra "expert", send a mail to "user2" with delivery report requierement.
- connect with "user2", open mail previously send with "user1" in zimbra "simple".
- connect back with "user1" and see that we have a delivery report in mail list.
- try to re-open same mail with "user2", and see that no more delivery report is send.

## Issue ticket number and link
[Jira - MZI-76](https://entsupport.gdapublic.fr/browse/MZI-76)

## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [x] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: zimbra/src/main/java/fr/openent/zimbra/controllers/ZimbraController.java [hunk 1]
```diff
@@ -406,13 +400,13 @@ public void returnMails(final HttpServerRequest request) {
     public void getReturnedMails(final HttpServerRequest request) {
         String structureId = request.params().get(ZIMBRA_ID_STRUCTURE);
         returnedMailService.getMailReturned(structureId, returnedMails -> {
```
Comment (qpic-cgi) [inline zimbra/src/main/java/fr/openent/zimbra/controllers/ZimbraController.java:402]: indent below

File: zimbra/src/main/java/fr/openent/zimbra/controllers/ZimbraController.java [hunk 2]
```diff
@@ -535,8 +529,8 @@ public void uploadAttachment(final HttpServerRequest request) {
             if (event.isRight()) {
                 JsonObject document = event.right().getValue();
                 String file = document.getString("file");
```
Comment (qpic-cgi) [inline zimbra/src/main/java/fr/openent/zimbra/controllers/ZimbraController.java:531]: indent below

File: zimbra/src/main/java/fr/openent/zimbra/helper/HttpClientHelper.java [hunk 3]
```diff
@@ -50,7 +50,7 @@ public static WebClientOptions getWebClientOptions() {
         if (System.getProperty("httpclient.proxyHost") != null) {
             options.setProxyOptions(HttpClientHelper.proxyOptions());
         }
-        int maxPoolSize = Zimbra.appConfig.getHttpClientMaxPoolSize();
+        int maxPoolSize = Zimbra.appConfig != null ? Zimbra.appConfig.getHttpClientMaxPoolSize() : 0;
```
Comment (qpic-cgi) [inline zimbra/src/main/java/fr/openent/zimbra/helper/HttpClientHelper.java:53]: edit for tests services (config is not necessary)

File: zimbra/src/main/java/fr/openent/zimbra/helper/ServiceManager.java [hunk 4]
```diff
@@ -77,43 +77,49 @@ public class ServiceManager {
...
+        JsonObject rawConfig = config != null ? config.getRawConfig() : null;
 
-        timelineHelper = new TimelineHelper(vertx, eb, rawConfig);
-        EmailFactory emailFactory = new EmailFactory(vertx, rawConfig);
-        emailSender = emailFactory.getSender();
+        if (rawConfig != null) {
+            timelineHelper = new TimelineHelper(vertx, eb, rawConfig);
+            EmailFactory emailFactory = new EmailFactory(vertx, rawConfig);
+            emailSender = emailFactory.getSender();
+        }
```
Comment (qpic-cgi) [inline zimbra/src/main/java/fr/openent/zimbra/helper/ServiceManager.java:92]: edit for tests services (config is not necessary)

File: zimbra/src/main/java/fr/openent/zimbra/helper/ServiceManager.java [hunk 5]
```diff
@@ -77,43 +77,49 @@ public class ServiceManager {
...
+            this.slackService = new SlackService(vertx, config.getSlackConfiguration());
+
+            this.sqlSynchroService = new SqlSynchroService(config.getDbSchema());
+            initDbMailService(config);
+            this.soapService = new SoapZimbraService(vertx, cacheService, slackService, config.getCircuitBreakerOptions());
+            this.sqlAddressBookService = new SqlAddressBookService(config.getDbSchema());
+            this.addressBookService = new AddressBookService(sqlAddressBookService);
+            this.userService = new UserService(soapService, synchroUserService, dbMailServiceApp,
+                    synchroAddressBookService, addressBookService, eb);
+        }
```
Comment (qpic-cgi) [inline zimbra/src/main/java/fr/openent/zimbra/helper/ServiceManager.java:109]: edit for tests services (config is not necessary)

File: zimbra/src/main/java/fr/openent/zimbra/helper/ServiceManager.java [hunk 6]
```diff
@@ -130,23 +136,27 @@ private ServiceManager(Vertx vertx, EventBus eb, String pathPrefix) {
...
 
     public static ServiceManager init(Vertx vertx, EventBus eb, String pathPrefix) {
-        if(serviceManager == null) {
-            serviceManager = new ServiceManager(vertx, eb, pathPrefix);
+        return init(vertx, eb, pathPrefix, Zimbra.appConfig);
+    }
+
+    public static ServiceManager init(Vertx vertx, EventBus eb, String pathPrefix, ConfigManager config) {
+        if (serviceManager == null) {
+            serviceManager = new ServiceManager(vertx, eb, pathPrefix, config);
```
Comment (qpic-cgi) [inline zimbra/src/main/java/fr/openent/zimbra/helper/ServiceManager.java:159]: edit for tests services (config is not necessary)

File: zimbra/src/main/java/fr/openent/zimbra/helper/ServiceManager.java [hunk 7]
```diff
@@ -267,11 +277,17 @@ public SynchroLauncher getSynchroLauncher() {
         return synchroLauncher;
     }
```
Comment (qpic-cgi) [inline zimbra/src/main/java/fr/openent/zimbra/helper/ServiceManager.java:279]: indent below

File: zimbra/src/main/java/fr/openent/zimbra/service/DbMailServiceFactory.java [hunk 8]
```diff
@@ -37,7 +37,7 @@ public DbMailServiceFactory(Vertx vertx, SqlSynchroService sqlSynchroService) {
     public DbMailService getDbMailService(String serviceName) {
         switch (serviceName) {
             case ConfigManager.SYNC_SQL:
-                return new SqlDbMailService(Zimbra.appConfig.getDbSchema());
+                return new SqlDbMailService(Zimbra.appConfig != null ? Zimbra.appConfig.getDbSchema() : "");
```
Comment (qpic-cgi) [inline zimbra/src/main/java/fr/openent/zimbra/service/DbMailServiceFactory.java:40]: edit for tests services (config is not necessary)

**Review by alicedraillard — APPROVED** [review]

**Review by BastienBenard — APPROVED** [review]

File: zimbra/src/main/resources/public/ts/controllers/controller.ts [hunk 9]
```diff
@@ -382,6 +393,7 @@ export let zimbraController = ng.controller("ZimbraController", [
             setCurrentMail(mail, true);
             try {
                 await mail.open(false,notMakeItRead);
+                await sendDeliveryReport(mail);
```
Comment (Sofianebernoussi) [inline zimbra/src/main/resources/public/ts/controllers/controller.ts:396]: mail.sendDeliveryReport() maybe

Comment (qpic-cgi) [reply zimbra/src/main/resources/public/ts/controllers/controller.ts:396]: et non ! regarde la ligne 201

File: zimbra/src/main/java/fr/openent/zimbra/controllers/ZimbraController.java [hunk 10]
```diff
@@ -1350,4 +1344,14 @@ public void checkIfIdGroup(HttpServerRequest request) {
         String idToCheck = request.params().get(MESSAGE_ID);
         userService.requestIfIdGroup(idToCheck, defaultResponseHandler(request));
     }
+
+
+    @Post("/message/:id/deliveryReport")
+    @ApiDoc("Send mail to acknowledge receipt to original sender")
+    public void sendDeliveryReport(HttpServerRequest request) {
+        getUserInfos(eb, request, user ->
```
Comment (sjollois) [inline zimbra/src/main/java/fr/openent/zimbra/controllers/ZimbraController.java:1352]: Dans le cas où le user est null, tu fais comment ?

**Review by sjollois — COMMENTED** [review]
petit commentaire

File: zimbra/src/main/java/fr/openent/zimbra/controllers/ZimbraController.java (hunk 10 above)
Comment (qpic-cgi) [reply zimbra/src/main/java/fr/openent/zimbra/controllers/ZimbraController.java:1352]: Bonne remarque, il manque le SecuredAction que j'ai oublié.
Je vais le rajouter tout de suite

**Pull request merged**